### PR TITLE
Weight bias init update

### LINKED ISSUE #85: SELU Incorrect weight and Bias initialization
https://github.com/Aifred-Health/Vulcan/blob/master/vulcanai/net.py#L327

ref: https://github.com/AlexiaJM/Deep-learning-with-cats/issues/11

### PR DESCRIPTION
**Pull request by joemehltretter**
Fix #85 
Created `selu_init_` function to mimic pytorch's method for initializing with a normal distribution based on standard deviation of the number of input features. In `DenseUnit` and `ConvUnit` class when checking if activation is SELU set the initilizer to `selu_int_` function and set `bias_init` to 0.0.

### CODE REVIEW

File: vulcanai2/models/layers.py [hunk 1]
```diff
@@ -62,9 +63,14 @@ def _init_bias(self):
         if self.bias_init is None, then pytorch default bias
         will be assigned to the kernel
         """
-        if self.bias_init:
+        if self.bias_init is not None:
```
Comment (rfratila) [inline vulcanai2/models/layers.py]: @joemehltretter I thought you were going to change it to passing the bias_init a function to use to initialize - the same way that the weight is initialized

File: vulcanai2/models/layers.py [hunk 2]
```diff
@@ -62,9 +63,14 @@ def _init_bias(self):
...
             nn.init.constant_(self._kernel.bias, self.bias_init)
 
+def selu_init_(tensor, mean=0):
```
Comment (rfratila) [inline vulcanai2/models/layers.py]: please add some documentation to this function

File: vulcanai2/models/layers.py [hunk 3]
```diff
@@ -121,16 +127,14 @@ def __init__(self, in_features, out_features,
                                         norm, dropout)
         self.in_features = in_features
         self.out_features = out_features
-
+        self.activation = activation
```
Comment (rfratila) [inline vulcanai2/models/layers.py]: why store it within the class attributes? I don't believe it is used outside of the init for Dense or Conv Unit

File: vulcanai2/models/layers.py [hunk 4]
```diff
@@ -146,6 +150,9 @@ def __init__(self, in_features, out_features,
         # Activation/Non-Linearity
         if activation is not None:
             self.add_module('_activation', activation)
+            if isinstance(activation, nn.SELU):
+                self.initializer = selu_init_
```
Comment (rfratila) [inline vulcanai2/models/layers.py]: can we call it `weight_init` instead of initializer so it matches the `bias_init` naming convention?

**Review by rfratila — CHANGES_REQUESTED** [review]

File: vulcanai2/models/layers.py (hunk 1 above)
Comment (joemehltretter) [reply vulcanai2/models/layers.py]: I must've forgot that, my bad. I'll do now.

File: vulcanai2/models/layers.py (hunk 3 above)
Comment (joemehltretter) [reply vulcanai2/models/layers.py]: That was something I added during testing and forgot to remove.

File: vulcanai2/models/layers.py [hunk 5]
```diff
@@ -63,14 +63,18 @@ def _init_bias(self):
...
 
-def selu_init_(tensor, mean=0):
+def selu_weight_init_(tensor, mean=0):
     with torch.no_grad():
         fan_in, _ = nn.init._calculate_fan_in_and_fan_out(tensor)
         std = math.sqrt(1. / fan_in)
-        return tensor.normal_(mean, std)
+        return nn.init.normal_(tensor.data, mean, std)
+
+def selu_bias_init_(tensor, const = 0.0):
```
Comment (rfratila) [inline vulcanai2/models/layers.py]: please remove the spaces for const default

File: vulcanai2/models/layers.py [hunk 6]
```diff
@@ -63,8 +64,17 @@ def _init_bias(self):
...
         """
         if self.bias_init:
-            nn.init.constant_(self._kernel.bias, self.bias_init)
+            self.bias_init(self._kernel.bias)
 
+def selu_weight_init_(tensor, mean=0):
+    with torch.no_grad():
+        fan_in, _ = nn.init._calculate_fan_in_and_fan_out(tensor)
+        std = math.sqrt(1. / fan_in)
+        return nn.init.normal_(tensor.data, mean, std)
```
Comment (rfratila) [inline vulcanai2/models/layers.py]: you don't need to get the `tensor.data`. that is a deprecated tool. just have it be `nn.init.normal_(tensor, mean, std)` the operation is already done in place

File: vulcanai2/models/layers.py [hunk 7]
```diff
@@ -63,8 +64,17 @@ def _init_bias(self):
...
+
+def selu_bias_init_(tensor, const = 0.0):
+    with torch.no_grad():
+        return nn.init.constant_(tensor.data, const)
```
Comment (rfratila) [inline vulcanai2/models/layers.py]: Same comment about turning `tensor.data` -> `tensor`

**Comment by rfratila** [comment]
Also @joemehltretter please update all the documentation with regards to the initializer -> weight_init name change and at the __init__ signatures

File: vulcanai2/models/layers.py (hunk 1 above)
Comment (rfratila) [reply vulcanai2/models/layers.py]: now that bias_init is a function, can you revert it back to `if self.bias_init`?

**Comment by joemehltretter** [comment]
> 
> 
> please also include the types to the parameters

@rfratila

File: vulcanai2/models/layers.py [hunk 8]
```diff
@@ -63,8 +64,48 @@ def _init_bias(self):
         will be assigned to the kernel
         """
         if self.bias_init:
-            nn.init.constant_(self._kernel.bias, self.bias_init)
+            self.bias_init(self._kernel.bias)
+
 
+def selu_weight_init_(tensor, mean=0.0):
```
Comment (rfratila) [inline vulcanai2/models/layers.py]: Can you put these in the `utils.py` file please @joemehltretter?

File: vulcanai2/models/layers.py [hunk 9]
```diff
@@ -52,8 +53,8 @@ def _init_weights(self):
         if self.initializer is None, then pytorch default weight
```
[the remark this answers is not in the document]
Comment (rfratila) [reply vulcanai2/models/layers.py]: @joemehltretter ^ also please update the example script since this changes the config dict required to build the layer ```dense_config = {
    'dense_units': [100, 50],
    'initializer': None,
    'bias_init': None,
    'norm': None,
    'dropout': 0.5,  # Single value or List
}```
should turn into:
dense_config = {
    'dense_units': [100, 50],
    'weight_init': None,
    'bias_init': None,
    'norm': None,
    'dropout': 0.5,  # Single value or List
}

File: vulcanai2/tests/models/test_layers.py [hunk 10]
```diff
@@ -1,42 +1,196 @@
...
+        )
+
+    def test_dense_selu_weight_change(self, dense_unit):
+        """Confirm SELU weight init properties hold for dense net."""
+        starting_weight = copy.deepcopy(dense_unit._kernel.weight)
+        std = round(math.sqrt(1. / 10), 1)
+
+        dense_unit.weight_init = selu_weight_init_
+        dense_unit._init_weights()
+        new_weight = copy.deepcopy(dense_unit._kernel.weight)
```
Comment (rfratila) [inline vulcanai2/tests/models/test_layers.py]: why do you need to copy? seems like it takes up more space than necessary.

File: vulcanai2/tests/models/test_layers.py [hunk 11]
```diff
@@ -1,42 +1,196 @@
...
+        assert (torch.equal(starting_weight, new_weight) is False)
+        assert pytest.approx(round(new_weight.std().item(), 1), 0.1) == std
```
Comment (rfratila) [inline vulcanai2/tests/models/test_layers.py]: What does calling `.item()` give you?

File: vulcanai2/tests/models/test_layers.py [hunk 12]
```diff
@@ -1,42 +1,196 @@
...
+        assert (int(new_weight.mean().item()) == 0.0)
+
+    def test_conv_selu_weight_change(self, conv_unit):
+        """Confirm SELU weight init properties hold for conv net."""
+        starting_weight = copy.deepcopy(conv_unit._kernel.weight)
+        std = round(math.sqrt(1. / 250), 1)
+        conv_unit.weight_init = selu_weight_init_
+        conv_unit._init_weights()
+        new_weight = copy.deepcopy(conv_unit._kernel.weight)
```
Comment (rfratila) [inline vulcanai2/tests/models/test_layers.py]: same comment on why the deep copy?

File: vulcanai2/tests/models/test_layers.py [hunk 13]
```diff
@@ -1,42 +1,196 @@
...
+        assert (round(new_bias.std().item(), 1) == 0.0)
+        assert (int(new_bias.mean().item()) == 0)
+
+    def test_conv_selu_bias_change(self, conv_unit):
+        """Confirm SELU bias init properties hold for conv net."""
+        starting_bias = copy.deepcopy(conv_unit._kernel.bias)
+
+        conv_unit.bias_init = selu_bias_init_
+        conv_unit._init_bias()
+        new_bias = copy.deepcopy(conv_unit._kernel.bias)
```
Comment (rfratila) [inline vulcanai2/tests/models/test_layers.py]: same comment on the deepcopy

File: vulcanai2/tests/models/test_layers.py [hunk 14]
```diff
@@ -1,42 +1,196 @@
...
+        assert (torch.equal(starting_bias, new_bias) is False)
+        assert (round(new_bias.std().item(), 1) == 0.0)
```
Comment (rfratila) [inline vulcanai2/tests/models/test_layers.py]: same comments for calling `.item()`

File: vulcanai2/tests/models/test_layers.py [hunk 15]
```diff
@@ -1,42 +1,196 @@
...
+        )
+
+    def test_selu_trained_dense(self, dnn_class):
+        """Confirm SELU weight and bias properties hold for a dense net."""
+        std = round(math.sqrt(1. / 200), 1)
+        test_input = torch.ones([10, *dnn_class.in_dim]).float()
+        test_target = torch.LongTensor([0, 1, 2, 3, 4, 5, 6, 7, 8, 9])
+        test_dataloader = DataLoader(TensorDataset(test_input, test_target))
+        dnn_class.fit(test_dataloader, test_dataloader, 5)
+        trained = copy.deepcopy(dnn_class.network.dense_0._kernel)
```
Comment (rfratila) [inline vulcanai2/tests/models/test_layers.py]: same comment on deepcopy

File: vulcanai2/tests/models/test_layers.py [hunk 16]
```diff
@@ -1,42 +1,196 @@
...
+        assert (round(trained.weight.std().item(), 1) == std)
```
Comment (rfratila) [inline vulcanai2/tests/models/test_layers.py]: sorry but same comment on `.item()`

File: vulcanai2/tests/models/test_layers.py [hunk 17]
```diff
@@ -1,42 +1,196 @@
...
+        assert (int(trained.weight.mean().item()) == 0.0)
+        assert (round(trained.bias.std().item(), 1) == 0.0)
+        assert (int(trained.bias.mean().item()) == 0.0)
+
+    def test_selu_trained_conv(self, cnn_class):
+        """Confirm SELU weight and bias properties hold for a conv net."""
+        std = round(math.sqrt(1. / 25), 1)
```
Comment (rfratila) [inline vulcanai2/tests/models/test_layers.py]: how did you calculate the fan_in for conv?

File: vulcanai2/tests/models/test_layers.py [hunk 18]
```diff
@@ -1,42 +1,196 @@
...
+        test_input = torch.ones([10, *cnn_class.in_dim]).float()
+        test_target = torch.LongTensor([0, 1, 2, 3, 4, 5, 6, 7, 8, 9])
+        test_dataloader = DataLoader(TensorDataset(test_input, test_target))
+        cnn_class.fit(test_dataloader, test_dataloader, 5)
+        trained = copy.copy(cnn_class.network.conv_0._kernel)
```
Comment (rfratila) [inline vulcanai2/tests/models/test_layers.py]: same comments on copy and `.item()`

File: vulcanai2/tests/models/test_layers.py (hunk 11 above)
Comment (joemehltretter) [reply vulcanai2/tests/models/test_layers.py]: @rfratila `new_weight.std()` is a tensor so `.item()` gives you the value itself.

File: vulcanai2/tests/models/test_layers.py (hunk 10 above)
Comment (joemehltretter) [reply vulcanai2/tests/models/test_layers.py]: @rfratila It's not necessary I will remove.

File: vulcanai2/tests/models/test_layers.py (hunk 17 above)
Comment (joemehltretter) [reply vulcanai2/tests/models/test_layers.py]: @rfratila I was doing it manually when setting up the tests. I can set it up to do it automatically if you'd like?

Comment (rfratila) [reply vulcanai2/tests/models/test_layers.py]: if possible having it done automatically would be preferred in case the fixture itself changes. thanks!

Comment (joemehltretter) [reply vulcanai2/tests/models/test_layers.py]: @rfratila Just pushed to automatically calculate

File: vulcanai2/tests/models/test_layers.py (hunk 15 above)
Comment (joemehltretter) [reply vulcanai2/tests/models/test_layers.py]: @rfratila Just pushed update that removed copy

File: vulcanai2/tests/models/test_layers.py (hunk 13 above)
Comment (joemehltretter) [reply vulcanai2/tests/models/test_layers.py]: @rfratila Just pushed update that removed copy

File: vulcanai2/tests/models/test_layers.py (hunk 12 above)
Comment (joemehltretter) [reply vulcanai2/tests/models/test_layers.py]: @rfratila Just pushed update that removed copy

File: vulcanai2/tests/models/test_layers.py (hunk 10 above)
Comment (joemehltretter) [reply vulcanai2/tests/models/test_layers.py]: @rfratila Just pushed update that removed copy

File: vulcanai2/tests/models/test_layers.py [hunk 19]
```diff
@@ -82,22 +82,24 @@ def conv_unit(self):
...
+        new_weight = dense_unit._kernel.weight
         assert (torch.equal(starting_weight, new_weight) is False)
         assert pytest.approx(round(new_weight.std().item(), 1), 0.1) == std
         assert (int(new_weight.mean().item()) == 0.0)
 
     def test_conv_selu_weight_change(self, conv_unit):
         """Confirm SELU weight init properties hold for conv net."""
         starting_weight = copy.deepcopy(conv_unit._kernel.weight)
-        std = round(math.sqrt(1. / 250), 1)
+        fan_in = conv_unit._kernel.in_channels * conv_unit._kernel.kernel_size[0] * conv_unit._kernel.kernel_size[1]
```
Comment (rfratila) [inline vulcanai2/tests/models/test_layers.py]: you can probably use the fan_in pytorch functionality?

File: vulcanai2/tests/models/test_layers.py (hunk 17 above)
Comment (rfratila) [reply vulcanai2/tests/models/test_layers.py]: Thanks @joemehltretter ! Just made one last comment but it should be good to go afterwards :)

File: vulcanai2/tests/models/test_layers.py (hunk 19 above)
Comment (rfratila) [reply vulcanai2/tests/models/test_layers.py]: or you could make it independent of the shape of the kernel (i.e. right now it will only work with 2D convo kernels)

Comment (joemehltretter) [reply vulcanai2/tests/models/test_layers.py]: @rfratila Made the update to be able to handle different sized kernels

File: vulcanai2/models/layers.py [hunk 20]
```diff
@@ -49,21 +50,21 @@ def _init_weights(self):
         """
         Initialize the weights.
 
-        if self.initializer is None, then pytorch default weight
-        will be assigned to the kernel
+        If self.weight_init is None, then pytorch default weight
```
Comment (priyatharsan) [inline vulcanai2/models/layers.py]: I think this is a typo: `If` -> `if`

File: vulcanai2/models/layers.py [hunk 21]
```diff
@@ -49,21 +50,21 @@ def _init_weights(self):
...
+        if self.weight_init:
+            self.weight_init(self._kernel.weight)
 
     def _init_bias(self):
         """
         Initialize the bias.
 
-        if self.bias_init is None, then pytorch default bias
-        will be assigned to the kernel
+        If self.bias_init is None, then pytorch default bias
```
Comment (priyatharsan) [inline vulcanai2/models/layers.py]: same: `If` -> `if`

**Review by priyatharsan — APPROVED** [review]

File: vulcanai2/models/layers.py (hunk 20 above)
Comment (rfratila) [reply vulcanai2/models/layers.py]: it's a comment so grammar applies

File: vulcanai2/models/layers.py (hunk 21 above)
Comment (rfratila) [reply vulcanai2/models/layers.py]: It's a comment so it's fine

**Review by rfratila — APPROVED** [review]

File: vulcanai2/models/utils.py [hunk 22]
```diff
@@ -174,34 +178,82 @@ def hook(module, input, output):
...
+    Returns
+    -------
+    torch.tensor
+        weight tensor with normal distribution
+
+    """
+    with torch.no_grad():
+        fan_in, _ = nn.init._calculate_fan_in_and_fan_out(tensor)
+        std = math.sqrt(1. / fan_in)
+        return nn.init.normal_(tensor, mean, std)
```
Comment (jcowles) [inline vulcanai2/models/utils.py:215]: This function seems identical to:
`nn.init.kaiming_normal_(mode="fan_in", nonlinearity="linear")`

Comment (rfratila) [reply vulcanai2/models/utils.py:215]: Hi @jcowles that's a great find! If you would like to submit a PR to update this, that would be fantastic.

cc @nirtiac
